working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+expressions.
+
+1. Initial constraints: Create some constraints on type parameters by trying to
+   match the return type of the invocation target as a subtype of the incoming
+   type context.  (For a constructor invocation, the return type of the
+   invocation target is considered the raw uninstantiated type of the class
+   enclosing the constructor declaration.)
+
+2. Downwards inference: partially solve the set of type constraints accumulated
+   in step 1, to produce a preliminary mapping of type parameters to type
+   schemas.
+   
+3. Visit arguments (legacy): If experiment flag `inference-update-1` is not
+   enabled, recursively perform inference on all arguments to the invocation in
+   source order.  Obtain the type contexts for the recursive inference by
+   substituting the preliminary mapping (from step 2) into the corresponding
+   parameter types of the invocation target.  For each argument that is
+   recursively inferred, create additional constraints on type parameters using
+   the resulting static type.  Then proceed to step 5.
+
+4. Visit arguments (new): If experiment flag `inference-update-1` is enabled,
+   topologically sort the arguments into stages (see [topological
+   sort](#Topological-sort) below), and then for each stage, do the following:
+
+   * Recursively infer all arguments selected for this stage that are not
+     function literals (see [what constitutes a function
+     literal?](#what-constitutes-a-function-literal), in source order.  (Note
+     that an invariant of the topological sort is that non-function literals are
+     always placed in the first stage).  The logic for obtaining type contexts
+     and for creating additional constraints is the same as in step 3.
+
+   * Recursively infer all arguments selected for this stage that are function
+     literals, in source order.
+
+   * Horizontal inference: if this is not the last stage, partially solve the
+     set of type constraints accumulated so far, to produce an updated
+     preliminary mapping of type parameters to type schemas.  Otherwise proceed
+     to step 5.
+
+5. Upwards inference: solve the set of type constraints accumulated so far, to
+   produce a final mapping of type parameters to types.  Check that each type is
+   a subtype of the bound of its corresponding type parameter.
+
+6. Type checking: Check that the static type of each argument is assignable to
+   the type obtained by substituting the final mapping (from step 5) into the
+   corresponding parameter type of the invocation target.
+
+7. Assign static type: Finally, obtain the static type of the invocation by
+   substituting the final mapping (from step 5) into the return type of the
+   invocation target.
+
+(Note that step 4 is the only step that differs from the inference implemented
+in Dart today, and it only takes effect if the experiment flag
+`inference-update-1` is enabled, hence this feature is backward compatible.)
```

Comment:
I wasn't really trying to claim that code that works without the flag enabled would continue to work with the flag enabled.  I was just trying to say that if the experiment flag is disabled, the implementations will continue to behave the way they've always behaved.  Which is kind of a vacuous statement anyhow, since the whole proposal is meant to only take effect when the experiment flag is enabled.  So I've just dropped this paragraph.